### ISSUE TITLE
The complications of dynamic role allocation

### ISSUE THREAD

**Issue by martinthomson** [private-attribution/ipa#376]
We assume that shares are allocated in a ring as follows:

* H1 gets s1, s2
* H2 gets s2, s3
* H3 gets s3, s1

However, clients and record collectors will identify helpers by their identity (for instance: A, B, C) and roles will be allocated by the leader.

If the leader allocates (A, B, C) -> (H1, H2, H3) or (B, C, A) or (C, A, B), then the order of allocations will ensure that the shares will be properly assigned.  These allocations preserve the order of helps so that A is to the left of B is to the left of C is to the left of A.

However, the other three possible allocations: (A, C, B), (B, A, C), and (C, B, A) reverse the ring order.  This works, but only if each helper knows that this is the case, so that it swaps the left and right values.

I think that we need something in the helper party network configuration that sets a canonical ring order, otherwise we'll potentially end up with a swapped order.  A failure here would either be disastrous for all of our security guarantees or would cause an immediate failure.  I'm not sure which: all of our algorithms make assumptions about how shares are replicated.

**Comment by richajaindce** [comment]
There are several subprotocols in sort which make this assumption. Can we tie the helper identity from report collection and let it remain the same throughout the computation?

**Comment by martinthomson** [comment]
Role allocation should be stable for an entire query - anything else is too hard to reason about.  The main point here is that we should either maintain ring order, or - if the order is reversed - also flip the two replicated shares.  I believe that is all that is necessary.

**Comment by akoshelev** [comment]
good catch, I can't think of any security implications, but the whole computation will likely result in garbage output. I feel like it is probably easier to start with canonical order and avoid "illegal" allocations. I suspect that we may want to reconsider that in the future because likely it will result in poor load balancing capabilities of the helper rings, but it should be easy to tweak later.
